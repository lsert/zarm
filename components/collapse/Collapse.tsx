import React, { PureComponent, Children, cloneElement, ReactElement } from 'react';
import classnames from 'classnames';
import { BaseCollapseProps } from './PropsType';
import { isArray } from '../utils/validate';

export interface CollapseProps extends BaseCollapseProps {
  prefixCls?: string;
  className?: string;
}

export default class Collapse extends PureComponent<CollapseProps, any> {
  static defaultProps = {
    prefixCls: 'za-collapse',
    multiple: false,
    animated: false,
    onChange: () => {},
  };

  static Item: any;

  constructor(props) {
    super(props);

    this.state = {
      activeKey: this.getActiveKey(props),
    };
  }

  componentWillReceiveProps(nextProps) {
    if (!this.isPropEqual(this.props.activeKey, nextProps.activeKey)) {
      this.setState({
        activeKey: this.getActiveKey(nextProps),
      });
    }
  }

  onItemChange = key => {
    if (!key) {
      return;
    }
    const { onChange, multiple } = this.props;
    const { activeKey } = this.state;
    const hasKey = activeKey.indexOf(key) > -1;
    let newActiveKey: Array<string> = [];
    if (multiple) {
      if (hasKey) {
        newActiveKey = activeKey.filter(i => i !== key);
      } else {
        newActiveKey = activeKey.slice(0);
        newActiveKey.push(key);
      }
    } else {
      newActiveKey = hasKey ? [] : [key];
    }
    this.setState({
      activeKey: newActiveKey,
    });
    onChange(key);
  }

  getActiveKey(props) {
    const { activeKey, defaultActiveKey, multiple } = props;

    const defaultKey = (activeKey || activeKey === 0) ? activeKey : defaultActiveKey;

    if (defaultKey || defaultKey === 0) {
      if (isArray(defaultKey)) {
        return !multiple ?
        [String(defaultKey[0])] : (defaultKey as Array<any>).map(key => String(key));
      } else {
        return [String(defaultKey)];
      }
    }

    return [];
  }

  isPropEqual(cur, next) {
    if (isArray(next) && isArray(cur)) {
      return next.length === cur.length && next.every((key, i) => key === cur[i]);
    }

    return cur === next;
  }

  renderItems() {
    const { animated } = this.props;
    const { activeKey } = this.state;

    return Children.map(this.props.children, (ele: any) => {
      const { disabled } = ele.props;
      const key = ele.key && String(ele.key);
      const isActive = activeKey.indexOf(key) > -1;
      return cloneElement(ele as ReactElement<any>, {
        animated,
        isActive,
        onItemChange: disabled ? null : () => this.onItemChange(key),
      });
    });
  }

  render() {
    const { prefixCls, className, style } = this.props;
    const cls = classnames(`${prefixCls}`, className);

    return (
      <div className={cls} style={style}>
        {this.renderItems()}
      </div>
    );
  }
}
