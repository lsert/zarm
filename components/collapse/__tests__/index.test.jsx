import React from 'react';
import { render, mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Collapse from '../index';

describe('Collapse', () => {
  const props = {};

  it('renders correctly', () => {
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with collapse mode', () => {
    props.multiple = true;
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with animated', () => {
    props.animated = true;
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with defaultActiveKey', () => {
    props.defaultActiveKey = [1];
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with string defaultActiveKey', () => {
    props.defaultActiveKey = '1';
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with number defaultActiveKey', () => {
    props.defaultActiveKey = 0;
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with activeKey', () => {
    props.activeKey = [0];
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with string activeKey', () => {
    props.activeKey = '1';
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with number activeKey', () => {
    props.activeKey = 0;
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly with dynamic activeKey', () => {
    props.activeKey = [0];
    const wrapper = shallow(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.setProps({ activeKey: 1 });

    expect(wrapper.state('activeKey')).toEqual(['1']);
  });

  it('renders correctly with dynamic array activeKey', () => {
    props.activeKey = [0];
    const wrapper = shallow(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.setProps({ activeKey: [1] });

    expect(wrapper.state('activeKey')).toEqual(['1']);
  });

  it('renders correctly with defaultActiveKey and activeKey', () => {
    props.defaultActiveKey = 0;
    props.activeKey = 1;
    const wrapper = render(
      <Collapse {...props}>
        <Collapse.Item className="item0" key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item className="item1" key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    expect(wrapper.find('.active').length).toBe(1);
  });

  it('click collapse item correctly', () => {
    props.onChange = jest.fn();
    const wrapper = mount(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐" {...props}>
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.find('.za-collapse-item-title').simulate('click');
    expect(props.onChange).toBeCalled();
  });

  it('click collapse item correctly with disabled mode', () => {
    props.onChange = jest.fn();
    const wrapper = mount(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐" disabled {...props}>
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.find('.za-collapse-item-title').simulate('click');
    expect(props.onChange).not.toBeCalled();
  });

  it('collapse items toggle correctly with animated', (done) => {
    props.animated = true;
    const wrapper = mount(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.find('.za-collapse-item-title').simulate('click');
    setTimeout(() => {
      expect(wrapper.find('.active').length).toBe(1);
      done();
    }, 0);
  });

  it('negative item toggle correctly with animated', (done) => {
    props.animated = true;
    const wrapper = mount(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.find('.za-collapse-item-title').simulate('click');
    setTimeout(() => {
      expect(wrapper.find('.active').length).toBe(1);
      done();
    }, 0);
  });

  it('active item toggle correctly with animated', (done) => {
    props.animated = true;
    props.activeKey = '0';
    const wrapper = mount(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.find('.za-collapse-item-title').simulate('click');
    setTimeout(() => {
      expect(wrapper.find('.active').length).toBe(0);
      done();
    }, 0);
  });

  it('collapse items toggle correctly without multiple mode', () => {
    props.multiple = false;
    props.activeKey = [1];
    props.onChange = jest.fn();
    const wrapper = mount(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.find('.za-collapse-item-title').at(0).simulate('click');
    expect(wrapper.find('.active').length).toBe(1);
  });

  it('collapse items toggle correctly with multiple mode', () => {
    props.multiple = true;
    props.activeKey = [0, 1];
    props.onChange = jest.fn();

    const wrapper = mount(
      <Collapse {...props}>
        <Collapse.Item key="0" title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
        <Collapse.Item key="1" title="100元套餐">
          <div>100元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.find('.za-collapse-item-title').at(1).simulate('click');
    expect(wrapper.find('.active').length).toBe(1);
  });

  it('click should not trigger callback without key', () => {
    props.onChange = jest.fn();
    const wrapper = mount(
      <Collapse {...props}>
        <Collapse.Item title="50元套餐">
          <div>50元套餐内容</div>
        </Collapse.Item>
      </Collapse>
    );
    wrapper.find('.za-collapse-item-title').simulate('click');
    expect(props.onChange).not.toBeCalled();
  });
});
