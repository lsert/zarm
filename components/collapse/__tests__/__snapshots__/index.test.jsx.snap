// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse renders correctly 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse renders correctly with activeKey 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item active"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse renders correctly with animated 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse renders correctly with collapse mode 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse renders correctly with defaultActiveKey 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-collapse-item active"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse renders correctly with number activeKey 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item active"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse renders correctly with number defaultActiveKey 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item active"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse renders correctly with string activeKey 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-collapse-item active"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Collapse renders correctly with string defaultActiveKey 1`] = `
<div
  class="za-collapse"
>
  <div
    class="za-collapse-item"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        50元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          50元套餐内容
        </div>
      </div>
    </div>
  </div>
  <div
    class="za-collapse-item active"
  >
    <div
      class="za-collapse-item-title"
    >
      <div>
        100元套餐
      </div>
      <div
        class="za-collapse-item-arrow"
      />
    </div>
    <div
      class="za-collapse-item-content za-collapse-item-content-anim"
    >
      <div
        class="za-collapse-item-content-inner"
      >
        <div>
          100元套餐内容
        </div>
      </div>
    </div>
  </div>
</div>
`;
